File: frontend/backend/app/schemas/lncrna_chipseq_overlap.py
"""
lncRNA-ChIP-seq Overlap Analysis Schemas

Pydantic models for lncRNA binding site overlap with ChIP-seq peaks API
"""
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, List
from decimal import Decimal


class OverlapFilters(BaseModel):
    """Query parameters for lncRNA-ChIP-seq overlap analysis"""

    lncrna_gene_id: Optional[int] = Field(None, description="Filter by specific lncRNA gene ID")
    target_gene_id: Optional[int] = Field(None, description="Filter by specific target gene ID")
    mark_type: Optional[str] = Field(None, description="Filter by mark type(s), comma-separated (e.g., 'H3K27me3,H3K4me3')")
    cell_type: Optional[str] = Field(None, description="Filter by cell type(s), comma-separated (e.g., 'K562,GM12878')")
    chromosome: Optional[str] = Field(None, description="Filter by chromosome (e.g., 'chr1')")
    min_overlap_length: Optional[int] = Field(None, ge=1, description="Minimum overlap length in bp")
    min_binding_affinity: Optional[float] = Field(None, ge=0, description="Minimum binding affinity score")
    min_peak_strength: Optional[float] = Field(None, ge=0, description="Minimum peak fold enrichment")
    max_qvalue: Optional[float] = Field(0.05, ge=0, le=1, description="Maximum Q-value (FDR) for peaks")
    page: int = Field(1, ge=1, description="Page number (starts from 1)")
    page_size: int = Field(100, ge=1, le=1000, description="Items per page")
    sort_by: Optional[str] = Field("binding_affinity", description="Sort field (binding_affinity, overlap_length, peak_fold_enrichment)")
    sort_order: Optional[str] = Field("desc", description="Sort order (asc or desc)")

    model_config = ConfigDict(from_attributes=True)


class OverlapResult(BaseModel):
    """Single overlap result between lncRNA binding site and ChIP-seq peak"""

    # Identifiers
    overlap_id: str = Field(..., description="Unique overlap identifier (format: reg_{regulation_id}_peak_{peak_id})")
    regulation_id: int = Field(..., description="Regulation record ID")

    # Gene information
    lncrna_gene_id: int = Field(..., description="lncRNA gene ID")
    lncrna_name: Optional[str] = Field(None, description="lncRNA gene name")
    target_gene_id: int = Field(..., description="Target gene ID")
    target_gene_name: Optional[str] = Field(None, description="Target gene name")

    # ChIP-seq experiment information
    mark_type: str = Field(..., description="Epigenetic mark type (e.g., H3K27me3)")
    mark_category: str = Field(..., description="Mark category (repressive, activating, etc.)")
    cell_type: str = Field(..., description="Cell type/line")

    # Genomic coordinates
    chromosome: str = Field(..., description="Chromosome")
    lncrna_binding_start: int = Field(..., description="lncRNA binding site start position")
    lncrna_binding_end: int = Field(..., description="lncRNA binding site end position")
    peak_start: int = Field(..., description="ChIP-seq peak start position")
    peak_end: int = Field(..., description="ChIP-seq peak end position")

    # Overlap region
    overlap_start: int = Field(..., description="Overlap region start position")
    overlap_end: int = Field(..., description="Overlap region end position")
    overlap_length: int = Field(..., description="Overlap length in bp")

    # Signal strength metrics
    binding_affinity: Decimal = Field(..., description="lncRNA binding affinity score")
    peak_fold_enrichment: Decimal = Field(..., description="ChIP-seq peak fold enrichment")
    peak_qvalue: Optional[Decimal] = Field(None, description="ChIP-seq peak Q-value (FDR)")

    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float}
    )


class OverlapResponse(BaseModel):
    """Paginated response for overlap query"""

    total: int = Field(..., description="Total number of matching overlaps")
    page: int = Field(..., description="Current page number")
    page_size: int = Field(..., description="Items per page")
    total_pages: int = Field(..., description="Total number of pages")
    items: List[OverlapResult] = Field(default=[], description="List of overlap results")
    default_filter_applied: bool = Field(
        default=False,
        description="True if default chromosome filter (chr22) was applied for performance optimization"
    )
    effective_chromosome: Optional[str] = Field(
        default=None,
        description="The chromosome filter actually used in the query (may differ from requested if default was applied)"
    )
    using_materialized_view: bool = Field(
        default=False,
        description="True if the optimized materialized view was used for this query (faster for large chromosomes)"
    )

    model_config = ConfigDict(from_attributes=True)

    @staticmethod
    def calculate_total_pages(total: int, page_size: int) -> int:
        """Calculate total pages"""
        return (total + page_size - 1) // page_size if page_size > 0 else 0


class MarkTypeStats(BaseModel):
    """Statistics breakdown by mark type"""
    mark_type: str = Field(..., description="Epigenetic mark type name")
    count: int = Field(..., description="Number of overlaps for this mark type")
    avg_strength: float = Field(..., description="Average binding affinity for this mark type")

    model_config = ConfigDict(from_attributes=True)


class CellTypeStats(BaseModel):
    """Statistics breakdown by cell type"""
    cell_type: str = Field(..., description="Cell type/line name")
    count: int = Field(..., description="Number of overlaps for this cell type")

    model_config = ConfigDict(from_attributes=True)


class OverlapStatistics(BaseModel):
    """Summary statistics for overlap analysis"""

    total_overlaps: int = Field(..., description="Total number of overlaps")
    unique_lncrnas: int = Field(..., description="Number of unique lncRNAs")
    unique_target_genes: int = Field(..., description="Number of unique target genes")
    unique_marks: int = Field(..., description="Number of unique mark types")
    unique_cell_types: int = Field(default=0, description="Number of unique cell types")
    avg_overlap_length: float = Field(..., description="Average overlap length in bp")
    avg_binding_affinity: float = Field(..., description="Average binding affinity")
    avg_peak_strength: float = Field(..., description="Average peak fold enrichment")
    by_mark_type: List[MarkTypeStats] = Field(default=[], description="Statistics breakdown by mark type")
    by_cell_type: List[CellTypeStats] = Field(default=[], description="Statistics breakdown by cell type")
    default_filter_applied: bool = Field(
        default=False,
        description="True if default chromosome filter (chr22) was applied for performance optimization"
    )
    effective_chromosome: Optional[str] = Field(
        default=None,
        description="The chromosome filter actually used in the query (may differ from requested if default was applied)"
    )

    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={Decimal: float}
    )


# ============================================================================
# Heatmap Schemas
# ============================================================================

class HeatmapCell(BaseModel):
    """Single cell in the heatmap matrix"""
    x: str = Field(..., description="X-axis value (mark_type or cell_type)")
    y: str = Field(..., description="Y-axis value (lncrna or target gene)")
    value: float = Field(..., description="Metric value for this cell")

    model_config = ConfigDict(from_attributes=True)


class OverlapHeatmapResponse(BaseModel):
    """Response schema for heatmap endpoint"""
    x_labels: List[str] = Field(..., description="List of X-axis labels")
    y_labels: List[str] = Field(..., description="List of Y-axis labels")
    data: List[HeatmapCell] = Field(..., description="Heatmap cell data")
    metric: str = Field(..., description="Metric used for values (count, avg_binding_affinity, total_overlap_length)")
    total_combinations: int = Field(..., description="Total possible X*Y combinations")
    valid_combinations: int = Field(..., description="Number of combinations with data")
    default_filter_applied: bool = Field(
        default=False,
        description="True if default chromosome filter (chr1) was applied for performance optimization"
    )
    effective_chromosome: Optional[str] = Field(
        default=None,
        description="The chromosome filter actually used in the query (may differ from requested if default was applied)"
    )

    model_config = ConfigDict(from_attributes=True)

File: frontend/web/src/types/lncRNAChIPSeqOverlap.ts
/**
 * lncRNA-ChIP-seq Overlap Analysis Types
 * Phase 3.0 - Task 1.3
 *
 * Type definitions for lncRNA binding sites overlapping with ChIP-seq peaks.
 * This allows analysis of epigenetic regulation at lncRNA binding locations.
 */

import type { MarkType } from './chipseq'

/**
 * Filter parameters for overlap query
 */
export interface OverlapFilters {
  /** lncRNA gene ID to filter by */
  lncrna_gene_id?: number
  /** Target gene ID to filter by */
  target_gene_id?: number
  /** Mark type(s) - comma-separated for multiple (e.g., "H3K27me3,H3K4me3") */
  mark_type?: string
  /** Cell type(s) - comma-separated for multiple (e.g., "K562,GM12878") */
  cell_type?: string
  /** Chromosome filter */
  chromosome?: string
  /** Minimum overlap length in base pairs */
  min_overlap_length?: number
  /** Minimum binding affinity score */
  min_binding_affinity?: number
  /** Minimum peak signal/fold enrichment */
  min_peak_strength?: number
  /** Maximum Q-value (FDR) threshold */
  max_qvalue?: number
  /** Page number for pagination */
  page?: number
  /** Items per page */
  page_size?: number
  /** Sort field */
  sort_by?: 'overlap_length' | 'binding_affinity' | 'peak_fold_enrichment' | 'peak_qvalue'
  /** Sort direction */
  sort_order?: 'asc' | 'desc'
}

/**
 * Single overlap result between lncRNA binding site and ChIP-seq peak
 */
export interface OverlapResult {
  /** Unique identifier for this overlap */
  overlap_id: string

  // lncRNA information
  /** lncRNA gene ID */
  lncrna_gene_id: number
  /** lncRNA gene name/symbol */
  lncrna_name: string | null

  // Target gene information
  /** Target gene ID */
  target_gene_id: number
  /** Target gene name/symbol */
  target_gene_name: string | null

  // ChIP-seq mark information
  /** Histone modification mark type */
  mark_type: MarkType
  /** Mark category (repressive, activating, etc.) */
  mark_category: 'repressive' | 'activating' | 'bivalent_component' | 'enhancer' | 'elongation' | 'other'
  /** Cell type/line where this was observed */
  cell_type: string

  // Genomic coordinates
  /** Chromosome */
  chromosome: string
  /** lncRNA binding site start */
  lncrna_binding_start: number
  /** lncRNA binding site end */
  lncrna_binding_end: number
  /** ChIP-seq peak start */
  peak_start: number
  /** ChIP-seq peak end */
  peak_end: number
  /** Overlap region start */
  overlap_start: number
  /** Overlap region end */
  overlap_end: number
  /** Overlap length in base pairs */
  overlap_length: number

  // Functional scores
  /** lncRNA-target binding affinity score */
  binding_affinity: number
  /** ChIP-seq peak fold enrichment */
  peak_fold_enrichment: number
  /** ChIP-seq peak Q-value (FDR) */
  peak_qvalue: number | null
}

/**
 * Paginated response for overlap queries
 */
export interface OverlapResponse {
  /** Total number of overlaps matching filters */
  total: number
  /** Current page number */
  page: number
  /** Items per page */
  page_size: number
  /** Array of overlap results for current page */
  items: OverlapResult[]
}

/**
 * Cursor (keyset) pagination response for overlap queries
 */
export interface OverlapCursorResponse {
  /** Total number of overlaps matching filters */
  total: number
  /** Items per page */
  page_size: number
  /** Array of overlap results for current page */
  items: OverlapResult[]
  /** Opaque cursor token for fetching the next page */
  next_cursor: string | null
  /** True if there are more results after this page */
  has_more: boolean
  /** True if default chromosome filter (chr22) was applied */
  default_filter_applied: boolean
  /** The chromosome filter actually used in the query */
  effective_chromosome: string | null
  /** True if the optimized materialized view was used */
  using_materialized_view: boolean
}

/**
 * Summary statistics for overlap analysis (Phase 2)
 */
export interface OverlapSummary {
  /** Total number of overlaps */
  total_overlaps: number
  /** Number of unique lncRNAs involved */
  unique_lncrnas: number
  /** Number of unique target genes involved */
  unique_target_genes: number
  /** Number of unique mark types */
  unique_marks: number
  /** Number of unique cell types */
  unique_cell_types: number

  // Aggregate statistics
  /** Average overlap length */
  avg_overlap_length: number
  /** Average binding affinity */
  avg_binding_affinity: number
  /** Average peak strength */
  avg_peak_strength: number

  // Breakdowns
  /** Statistics by mark type */
  by_mark_type: Array<{
    mark_type: string
    count: number
    avg_strength: number
  }>
  /** Statistics by cell type */
  by_cell_type: Array<{
    cell_type: string
    count: number
  }>

  /** True if default chromosome filter (chr22) was applied for performance optimization */
  default_filter_applied?: boolean
  /** The chromosome filter actually used in the query */
  effective_chromosome?: string | null
}

export interface OverlapCrossSpeciesComparisonResponse {
  lncrna_core_id: number
  target_core_id: number | null
  species_names: Record<string, string>
  species_stats: Record<
    number,
    {
      species_id: number
      species_name: string
      lncrna_gene_id: number | null
      target_gene_id: number | null
      statistics: OverlapSummary
    }
  >
}

/**
 * Mark category mapping (aligned with ChIP-seq schema)
 */
export type MarkCategory = 'repressive' | 'activating' | 'bivalent_component' | 'enhancer' | 'elongation' | 'other' | 'open_chromatin'

/**
 * Cell type options (common ENCODE cell lines)
 */
export const CELL_TYPE_OPTIONS = [
  'K562',      // Chronic myelogenous leukemia
  'GM12878',   // Lymphoblastoid
  'HepG2',     // Hepatocellular carcinoma
  'H1-hESC',   // Human embryonic stem cells
  'HeLa-S3',   // Cervical cancer
  'A549',      // Lung carcinoma
  'MCF-7',     // Breast cancer
  'HMEC',      // Human mammary epithelial cells
] as const

/**
 * Chromosome options
 */
export const CHROMOSOME_OPTIONS = [
  'chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6',
  'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12',
  'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18',
  'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM'
] as const

export type CellType = typeof CELL_TYPE_OPTIONS[number]
export type Chromosome = typeof CHROMOSOME_OPTIONS[number]

/**
 * Heatmap data structure for overlap visualization (Phase 3.0 Phase 2)
 */
export interface OverlapHeatmapData {
  /** X-axis labels (mark types or cell types) */
  x_labels: string[]
  /** Y-axis labels (lncRNAs or target genes) */
  y_labels: string[]
  /** Heatmap data points */
  data: Array<{ x: string; y: string; value: number }>
  /** Current metric being displayed */
  metric: OverlapHeatmapMetric
  /** Total possible combinations */
  total_combinations: number
  /** Combinations with valid data */
  valid_combinations: number
}

/**
 * Heatmap metric options
 */
export type OverlapHeatmapMetric = 'count' | 'avg_binding_affinity' | 'total_overlap_length'

/**
 * X-axis options for heatmap
 */
export type OverlapHeatmapXAxis = 'mark_type' | 'cell_type'

/**
 * Y-axis options for heatmap
 */
export type OverlapHeatmapYAxis = 'lncrna' | 'target_gene'

/**
 * Heatmap request parameters
 */
export interface OverlapHeatmapParams {
  /** X-axis dimension */
  x_axis: OverlapHeatmapXAxis
  /** Y-axis dimension */
  y_axis: OverlapHeatmapYAxis
  /** Metric to display */
  metric: OverlapHeatmapMetric
  /** Top N items to include */
  top_n?: number
  /** Chromosome filter */
  chromosome?: string
  /** Minimum binding affinity filter */
  min_binding_affinity?: number
  /** Maximum Q-value filter */
  max_qvalue?: number
}
